composer/trainer/trainer.py

```diff
@@ -2848,7 +2852,20 @@ def _train_microbatches(
                         optimizer.zero_grad()
 
             # Tracker for gradient accumulation
-            current_batch_size = sum([self._train_data_spec.get_num_samples_in_batch(batch) for batch in microbatches])
+            if self.accumulate_train_batch_on_tokens:
+                current_batch_size = sum([self._train_data_spec.get_num_tokens_in_batch(b) for b in microbatches])
+                if current_batch_size == 0:
+                    warnings.warn(
+                        'Requested loss accumulation based on number of tokens in training batch, but zero tokens '
+                        'found (perhaps due to an improper DataSpec).',
+                    )
```

Comment:
@dakinggg  shoudl this be an error instead? eg u get all pad tokens? Maybe you need to do this if your dataset is uneven size and you don't want to `drop_last` but using FSDP so all ranks participate?

Reply:
Happy to emit an exception instead. I chose a warning initially based on https://github.com/mosaicml/composer/pull/3273#discussion_r1600672811

Reply:
exception seems reasonable to me

Reply:
Updated accordingly.